### PR TITLE
Expose HostingEngine.CreateApplicationServices

### PR DESCRIPTION
**Pull request by HaoK**
@davidfowl @divega Thoughts?  Basically this will be the replacement for the old HostingServices.Create for how to create ApplicationServices without having to go through the whole HostingEngine/context setup.  This eliminates the need for implementing IServerFactory just to create an ApplicationServices.

### CODE REVIEW

**Comment by HaoK** [comment]
So maybe we need to further tweak a bit, as we still need to short circuit the StartupLoader logic, see the matching SignalR change to use this method

https://github.com/aspnet/SignalR-Server/pull/89

**Comment by HaoK** [comment]
Tweaked it a bit, moved default logic which adds no-op startup methods when calling CreateApplicationServices

**Comment by HaoK** [comment]
Next question is if HostingContext is really the right input parameter... feels a bit too heavy still...

**Comment by HaoK** [comment]
Added support for automatically replacing ApplicationName and BasePath, which is mainly what the tests were doing

File: src/Microsoft.AspNet.Hosting/HostingContext.cs [hunk 1]
```diff
@@ -17,7 +17,9 @@ public class HostingContext
         public IApplicationBuilder Builder { get; set; }
 
         public string ApplicationName { get; set; }
-        public string WebRootPath { get; set; }
+        // REVIEW: BasePath and WebRootPath are basically setting same thing?
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingContext.cs]: isn't one the project directory and the other wwwroot?

File: src/Microsoft.AspNet.Hosting/HostingEngine.cs [hunk 2]
```diff
@@ -21,7 +23,7 @@ public class HostingEngine
 
         private readonly IServiceProvider _fallbackServices;
         private readonly ApplicationLifetime _appLifetime;
-        private readonly IApplicationEnvironment _applicationEnvironment;
+        private readonly IApplicationEnvironment _appEnv;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingEngine.cs]: why switch from readable name to a confusing one?

File: src/Microsoft.AspNet.Hosting/HostingEngine.cs [hunk 3]
```diff
@@ -29,18 +31,39 @@ public class HostingEngine
 
         public HostingEngine() : this(fallbackServices: null) { }
 
-        public HostingEngine(IServiceProvider fallbackServices)
+        public HostingEngine(IServiceProvider fallbackServices) : this(fallbackServices, applicationName: null, applicationPath: null) { }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingEngine.cs]: nit: long line

File: src/Microsoft.AspNet.Hosting/HostingEngine.cs [hunk 4]
```diff
@@ -29,18 +31,39 @@ public class HostingEngine
...
             _fallbackServices = new WrappingServiceProvider(_fallbackServices, _hostingEnvironment, _appLifetime);
         }
 
+        public static IServiceProvider CreateApplicationServices(HostingContext context)
+        {
+            return CreateApplicationServices(context, fallbackServices: null);
+        }
+
+        // REVIEW: so ugly ApplicationName used here as input
+        public static IServiceProvider CreateApplicationServices(HostingContext context, IServiceProvider fallbackServices)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingEngine.cs]: nit long line

File: src/Microsoft.AspNet.Hosting/HostingEngine.cs [hunk 5]
```diff
@@ -29,18 +31,39 @@ public class HostingEngine
...
+            _hostingEnvironment = new HostingEnvironment(_appEnv);
             _fallbackServices = new WrappingServiceProvider(_fallbackServices, _hostingEnvironment, _appLifetime);
         }
 
+        public static IServiceProvider CreateApplicationServices(HostingContext context)
+        {
+            return CreateApplicationServices(context, fallbackServices: null);
+        }
+
+        // REVIEW: so ugly ApplicationName used here as input
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingEngine.cs]: do you mean "This method exists for use in functional tests, not from product code"?

File: src/Microsoft.AspNet.Hosting/HostingEngine.cs [hunk 6]
```diff
@@ -29,18 +31,39 @@ public class HostingEngine
...
             _appLifetime = new ApplicationLifetime();
-            _applicationEnvironment = _fallbackServices.GetRequiredService<IApplicationEnvironment>();
-            _hostingEnvironment = new HostingEnvironment(_applicationEnvironment);
+            var appEnv = _fallbackServices.GetRequiredService<IApplicationEnvironment>();
+            _appEnv = new HostingApplicationEnvironment(appEnv, applicationPath, applicationName);
+            _hostingEnvironment = new HostingEnvironment(_appEnv);
             _fallbackServices = new WrappingServiceProvider(_fallbackServices, _hostingEnvironment, _appLifetime);
         }
 
+        public static IServiceProvider CreateApplicationServices(HostingContext context)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingEngine.cs]: what is the use case for this method?

File: src/Microsoft.AspNet.Hosting/HostingEngine.cs [hunk 7]
```diff
@@ -252,6 +273,51 @@ public object GetService(Type serviceType)
             }
         }
 
+        // Represents an application environment that overrides the base path of the original
+        // application environment in order to make it point to the test application folder.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingEngine.cs]: comment is misleading. you're creating one of these for the mainline scenario as well, just with `appPath` and `appName` both `null`.

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 8]
```diff
@@ -25,22 +25,32 @@ public class TestServer : IServerFactory, IDisposable
...
         }
 
+        // REVIEW: we can configure services via AppStartup or via hostContext.Services
+        public TestServer(IConfiguration config, IServiceProvider serviceProvider, Action<IApplicationBuilder> configureApp, ConfigureServicesDelegate configureServices)
+            : this(serviceProvider, new HostingContext()
+                {
+                    ApplicationName = "Test App",
+                    Configuration = config,
+                    StartupMethods = new StartupMethods(configureApp, configureServices)
+                })
```
Comment (dougbu) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: nit: ick.  how about a static method that returns the `HostingContext`?

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 9]
```diff
@@ -25,22 +25,32 @@ public class TestServer : IServerFactory, IDisposable
...
+        { }
+
         public Uri BaseAddress { get; set; } = new Uri("http://localhost/");
 
+        public static TestServer Create(HostingContext context)
```
Comment (dougbu) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: why does the test code now need to know `HostingContext`?  seems like moving the faked (but easily-created and understood) `IApplicationEnvironment` from test code into hosting has increased the concept count in code consuming these APIs.  would prefer to see just another `Create` method that takes the application name and path as parameters.

File: src/Microsoft.AspNet.Hosting/HostingContext.cs [hunk 10]
```diff
@@ -17,7 +17,9 @@ public class HostingContext
...
+        public string ApplicationBasePath { get; set; }
+        //public string WebRootPath { get; set; }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Hosting/HostingContext.cs]: actually, not sure why this class is changing at all.  seems unrelated to avoiding creation of `IApplicationEnvironment` implementations in test code.

**Comment by dougbu** [comment]
:watch:

**Comment by HaoK** [comment]
Replacing with PR with next iteration of hosting https://github.com/aspnet/Hosting/issues/204
